devlib/trace/dmesg.py

```diff
@@ -0,0 +1,185 @@
+
+    :param msg: Content of the entry
+    :type msg: str
+    """
+    def __init__(self, facility, level, ts, msg):
+        self.facility = facility
+        self.level = level
+        self.ts = ts
+        self.msg = msg
+
+    @classmethod
+    def from_str(cls, line):
+        """
+        Parses a "dmesg --decode" output line, formatted as following:
+        kern  :err   : [3618282.310743] nouveau 0000:01:00.0: systemd-logind[988]: nv50cal_space: -16
+
+        Or the more basic output:
+        [3618282.310743] nouveau 0000:01:00.0: systemd-logind[988]: nv50cal_space: -16
+
+        """
+
+        def parse_ts_msg(line):
+            ts_msg_regex = r'\[(.*?)\] (.*)'
+            match = re.match(ts_msg_regex, line.strip())
+            if not match:
+                raise ValueError('dmesg entry format not recognized: {}'.format(line))
+            ts, msg = match.groups()
+            return ts, msg
+
+        # If we can parse the timestamp directly, that is a basic line
+        try:
+            ts, msg = parse_ts_msg(line)
+        except ValueError:
+            facility, level, remainder = line.split(':', 2)
+            ts, msg = parse_ts_msg(remainder)
+            facility = facility.strip()
+            level = level.strip()
+        else:
+            facility = None
+            level = None
+
+        return cls(
+            facility=facility,
+            level=level,
+            ts=timedelta(seconds=float(ts.strip())),
+            msg=msg.strip(),
+        )
+
+    def __str__(self):
+        if self.facility and self.level:
+            prefix = '{facility}:{level}:'.format(
+                facility=self.facility,
+                level=self.level,
+            )
+        else:
+            prefix = ''
+
+        return '{prefix}[{ts}] {msg}'.format(
+            ts=self.ts.total_seconds(),
+            msg=self.msg,
+            prefix=prefix,
+        )
+
+
+class DmesgCollector(TraceCollector):
+    """
+    Dmesg output collector.
+
+    :param level: Minimum log level to enable. All levels that are more
+        critical will be collected as well.
+    :type level: str
+
+    :param facility: Facility to record, see dmesg --help for the list.
+    :type level: str
+
+    .. warning:: If BusyBox dmesg is used, facility and level will be ignored,
+        and the parsed entries will also lack that information.
+    """
+
+    # taken from "dmesg --help"
+    # This list needs to be ordered by priority
+    LOG_LEVELS = [
+        "emerg",        # system is unusable
+        "alert",        # action must be taken immediately
+        "crit",         # critical conditions
+        "err",          # error conditions
+        "warn",         # warning conditions
+        "notice",       # normal but significant condition
+        "info",         # informational
+        "debug",        # debug-level messages
+    ]
+
+    def __init__(self, target, level=LOG_LEVELS[-1], facility='kern'):
```

Comment:
Could we have a check for that facility being available (other than the `dmesg` exit code check) ?

I don't know how much you can rely on the output of `dmesg -h` but you could parse facilities (and log levels BTW) from there. Sort of something like this (but as a static/classmethod maybe) https://github.com/ARM-software/devlib/blob/master/devlib/trace/systrace.py#L61

Reply:
Here is the busybox's dmesg help :-)

```
# dmesg -h
dmesg: invalid option -- 'h'
BusyBox v1.29.2 (2018-12-18 16:55:05 GMT) multi-call binary.

Usage: dmesg [-c] [-n LEVEL] [-s SIZE]

Print or control the kernel ring buffer

	-c		Clear ring buffer after printing
	-n LEVEL	Set console logging level
	-s SIZE		Buffer size
	-r		Print raw message buffer
```

Note that there isn't even a `--help` option, so we can safely forget about any "introspection" capabilities being always available.

There is also relatively little value in practice, since 99% of devlib users will probably restrict that to `kern`. I had to get the list of log levels for another reason: util-linux `dmesg` wants an explicit list of all levels to show, i.e. it will not show "levels above X", it will show "level X". Since we usually want the "levels above X" behavior, I had to get the ordered list of levels.

Reply:
Fair enough!